_posts/2020-11-03-torch-tabular/torch_tabular.Rmd

```diff
@@ -0,0 +1,445 @@
+
+```{r}
+cardinalities <- map(
+  mushroom_data[ , 2:ncol(mushroom_data)], compose(nlevels, as.factor)) %>%
+  keep(function(x) x > 2) %>%
+  unlist() %>%
+  unname()
+
+num_numerical <- ncol(mushroom_data) - length(cardinalities) - 1
+
+fc1_dim <- 16
+fc2_dim <- 16
+
+model <- net(
+  cardinalities,
+  num_numerical,
+  fc1_dim,
+  fc2_dim
+)
+
+device <- if (cuda_is_available()) torch_device("cuda:0") else "cpu"
+
+model <- model$to(device = device)
+```
+
+## Training
+
+The training loop now is "business as usual":
+
+```{r}
+optimizer <- optim_adam(model$parameters, lr = 0.1)
+
+for (epoch in 1:20) {
+
+  model$train()
+  train_losses <- c()  
+
+  for (b in enumerate(train_dl)) {
+    optimizer$zero_grad()
+    output <- model(b$x[[1]]$to(device = device), b$x[[2]]$to(device = device))
+    loss <- nnf_binary_cross_entropy(output, b$y$to(dtype = torch_float(), device = device))
+    loss$backward()
+    optimizer$step()
+    train_losses <- c(train_losses, loss$item())
+  }
+
+  model$eval()
+  valid_losses <- c()
+
+  for (b in enumerate(valid_dl)) {
+    output <- model(b$x[[1]]$to(device = device), b$x[[2]]$to(device = device))
+    loss <- nnf_binary_cross_entropy(output, b$y$to(dtype = torch_float(), device = device))
+    valid_losses <- c(valid_losses, loss$item())
+  }
+
+  cat(sprintf("Loss at epoch %d: training: %3f, validation: %3f\n", epoch, mean(train_losses), mean(valid_losses)))
+}
+
+```
+
+    Loss at epoch 1: training: 0.274634, validation: 0.111689
+    Loss at epoch 2: training: 0.057177, validation: 0.036074
+    Loss at epoch 3: training: 0.025018, validation: 0.016698
+    Loss at epoch 4: training: 0.010819, validation: 0.010996
+    Loss at epoch 5: training: 0.005467, validation: 0.002849
+    Loss at epoch 6: training: 0.002026, validation: 0.000959
+    Loss at epoch 7: training: 0.000458, validation: 0.000282
+    Loss at epoch 8: training: 0.000231, validation: 0.000190
+    Loss at epoch 9: training: 0.000172, validation: 0.000144
+    Loss at epoch 10: training: 0.000120, validation: 0.000110
+    Loss at epoch 11: training: 0.000098, validation: 0.000090
+    Loss at epoch 12: training: 0.000079, validation: 0.000074
+    Loss at epoch 13: training: 0.000066, validation: 0.000064
+    Loss at epoch 14: training: 0.000058, validation: 0.000055
+    Loss at epoch 15: training: 0.000052, validation: 0.000048
+    Loss at epoch 16: training: 0.000043, validation: 0.000042
+    Loss at epoch 17: training: 0.000038, validation: 0.000038
+    Loss at epoch 18: training: 0.000034, validation: 0.000034
+    Loss at epoch 19: training: 0.000032, validation: 0.000031
+    Loss at epoch 20: training: 0.000028, validation: 0.000027
+
+While loss on the validation set is still decreasing, we'll soon see that this is good enough for 100% accuracy.
```

Comment:
Not sure what "this" is in the sentence above (this meaning 20 epochs?)

Reply:
right, made this a bit clearer